### PR TITLE
FSDP-2 StormScope 

### PR DESCRIPTION
**Pull request by negin513**
# PhysicsNeMo Pull Request
Migrates StormScope off FSDP1 onto FSDP2 (`fully_shard` / `FSDPModule`)....

## Description
FSDP1's flat-param machinery doesn't compose with `ShardTensor` / DTensor. The refactored `ShardTensor` in #1556 breaks FSDP1's backward pass in StormCast, so domain parallelism was not working with FSDP (only DDP).

**Fun fact:** StormCast was using `FSDP(model, sharding_strategy=NO_SHARD)` — which is effectively DDP with extra overhead. Moving to FSDP2 (`fully_shard`) gives us real sharding capability and DTensor compatibility.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [x] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Dependencies
- #1556 (ShardTensor refactor) must merge first

## Tests Added
**`examples/weather/stormcast/test_training.py`**
- `test_contiguity_channels_last` — verifies contiguity fix preserves channels-last activations (cuDNN fast path)
**`test/utils/test_checkpoint_distributed.py`** (4 new multi-GPU tests)

| Test | Validates |
|------|-----------|
| `test_fsdp2_checkpoint_roundtrip` | Basic `fully_shard` model+optimizer+scheduler save/load |
| `test_fsdp2_checkpoint_channels_last_roundtrip` | channels_last conv + contiguity fix + cross-rank parity |
| `test_fsdp2_shard_tensor_checkpoint_roundtrip` | 2-D mesh (ddp × domain) with ShardTensor (4-GPU) |
| `test_fsdp2_grad_scaler_checkpoint` | GradScaler state under FSDP2 + amp autocast |

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: examples/weather/stormcast/utils/parallel.py [hunk 1]
```diff
@@ -246,24 +251,29 @@ def distribute_model(self, model: torch.nn.Module) -> FSDP:
...
-        return FSDP(
-            model,
-            device_mesh=self.mesh["ddp"],
-            use_orig_params=False,  # Required for use with ShardTensor
-            sharding_strategy=ShardingStrategy.NO_SHARD,
-            sync_module_states=True,  # Ensure initialized weights match across ranks
-            forward_prefetch=True,  # Optimization for faster training
-            backward_prefetch=BackwardPrefetch.BACKWARD_PRE,  # Backward prefetching for overlap
-        )
+        # FSDP2 rejects non-contiguous parameters (PyTorch <= 2.10):
```
Comment (negin513) [inline examples/weather/stormcast/utils/parallel.py]: NOTE: this block exists solely for backward compatibility with PyTorch <= 2.10. Do we care about backward compatibility?

**Comment by negin513** [comment]
Flagging the changes not necessary once we merge #1556 (checkpoint + DTensor reliance).

**Comment by negin513** [comment]
Also if we drop FSDP1 support; how much that would simplify the checkpointing.

File: examples/weather/stormcast/test_training.py [hunk 2]
```diff
@@ -592,3 +585,55 @@ def test_model_types(
...
+    # 2. Channels-last input should still produce channels-last output
+    x = torch.randn(2, 3, 16, 16, device=device).to(memory_format=torch.channels_last)
+    assert x.is_contiguous(memory_format=torch.channels_last)
+
+    with torch.no_grad():
+        y = model(x)
+
+    assert y.is_contiguous(memory_format=torch.channels_last), (
+        "Output activations lost channels_last layout — cuDNN fast path is broken"
+    )
```
Comment (pzharrington) [inline examples/weather/stormcast/test_training.py]: I appreciate this test as a "show-your-work, Claude" confirmation that the `.contiguous()` call is acting as intended, but it's just testing an isolated and independent model (it'd be more of a unit test for pytorch itself or something). I suggest dropping it just to keep this file explicitly related to direct testing of recipe functionality, it's not adding much value to the actual test suite here which is supposed to catch bugs/regressions from changes to the recipe.

File: physicsnemo/utils/checkpoint.py [hunk 3]
```diff
@@ -165,14 +225,48 @@ def _has_non_fsdp_dtensors(
...
-    DTensors on its own mesh.  ``broadcast_from_rank0`` handles these
-    natively, so manual redistribution should be skipped.  Only
-    user-created DTensors (e.g. ShardTensor on a separate domain mesh)
-    require explicit redistribution.
+    """Return ``True`` when *dtensor_plc* requires the manual broadcast path.
+
+    The function answers a single operational question: does loading this
+    model's state dict need to bypass DCP's ``broadcast_from_rank0`` in
+    favour of the explicit ``broadcast_object_list`` /
+    ``_redistribute_sd_for_dtensor`` path?  The answer is ``True`` for:
```
Comment (pzharrington) [inline physicsnemo/utils/checkpoint.py]: Based on this monster docstring I can tell debugging this was fun 😅 given the more subtle meaning/varied cases addressed by this helper function, we should rename it to more clearly reflect this "single operational question". Maybe `_needs_dcp_broadcast_bypass`?

File: test/utils/test_checkpoint_distributed.py
Comment (pzharrington) [inline test/utils/test_checkpoint_distributed.py:1]: Good job adding some explicit FSDP2 tests in here. They are covering the core/critical functionality but I think it would be nice to expand and cover more of the other edge cases (e.g. `test_cross_mode_channels_last_model_load`, `test_load_model_weights_fsdp`, `test_load_model_weights_fsdp_shard_tensor`) for FSDP2. That will make it easier to transition from supporting FSDP1 to FSDP2 with confidence -- basically the ideal would be any test we have with FSDP is also covered for FSDP2. Then we can nuke all the FSDP1 tests after this PR lands 😈

File: examples/weather/stormcast/test_training.py (hunk 2 above)
Comment (negin513) [reply examples/weather/stormcast/test_training.py]: Haha!! It seems like I forgot to clean this up. Done.

File: physicsnemo/utils/checkpoint.py (hunk 3 above)
Comment (negin513) [reply physicsnemo/utils/checkpoint.py]: Fair point!! `_has_non_fsdp_dtensors` is a lie (such a bad name!).
I'll rename it to `_needs_dcp_broadcast_bypass` and tighten the docstring now that the name carries "the operational question".

**Review by pzharrington — APPROVED** [review]
All multigpu tests are passing for me locally. Great work! Next up will be ripping out all the FSDP1 stuff :)

**Comment by negin513** [comment]
This is ready to be merged. :-) 

I added some docstrings to make removal of FSDP1 support much easier later. (Here is [the issue](https://github.com/NVIDIA/physicsnemo-roadmap/issues/2611))


Related issue: Issue #148831: https://github.com/pytorch/pytorch/issues/148831

File: test/diffusion/test_domain_parallel_sampling.py [hunk 4]
```diff
@@ -71,7 +71,7 @@ def test_wrapper_timesteps_replicated(distributed_mesh):
     from physicsnemo.diffusion.noise_schedulers import DomainParallelNoiseScheduler
 
     scheduler = EDMNoiseScheduler()
-    wrapper = DomainParallelNoiseScheduler(scheduler, distributed_mesh)
+    wrapper = DomainParallelNoiseScheduler(scheduler, distributed_mesh, shard_dim=2)
```
Comment (CharlelieLrt) [inline test/diffusion/test_domain_parallel_sampling.py:74]: Why changing this test?

File: physicsnemo/utils/checkpoint.py [hunk 5]
```diff
@@ -94,12 +103,35 @@ def _unwrap_ddp_compile(
 
 
 def _unwrap_fsdp(model: torch.nn.Module) -> torch.nn.Module:
-    """Unwrap one FSDP layer (if present) to reach the user module."""
+    """
+    Unwrap one FSDP layer (if present) to reach the user module.
+    FSDP1 only.
```
Comment (CharlelieLrt) [inline physicsnemo/utils/checkpoint.py:108]: Is there a plan to drop support for FSDP1 later? These checkpoint loading utilities got significantly more and more complex recently

File: physicsnemo/utils/checkpoint.py [hunk 6]
```diff
@@ -577,15 +678,13 @@ def _unique_model_names(
...
+    """
+    Map a list of models to unique names derived from their class names.
 
     DDP and ``torch.compile`` wrappers are stripped, but FSDP wrappers are
     preserved so that the returned modules can be passed to PyTorch's DCP
     state-dict helpers when needed.
 
-    When multiple models share a class name a numeric suffix is appended
-    (e.g. ``"MyModel0"``, ``"MyModel1"``).
-
```
Comment (CharlelieLrt) [inline physicsnemo/utils/checkpoint.py:588]: I think it still does that, no?

File: physicsnemo/utils/checkpoint.py [hunk 7]
```diff
@@ -1364,12 +1520,9 @@ def get_checkpoint_dir(base_dir: Path | str, model_name: str) -> str:
...
 
 def _cache_if_needed(path: str) -> str:
-    r"""Return a local path for ``path``, downloading to cache if remote.
+    """
+    For remote fsspec URIs, downloads to a local cache and returns the local path.
 
-    For the ``"file"`` protocol the path is returned unchanged.  For remote
-    protocols the file is fetched via
-    :func:`~physicsnemo.core.filesystem._download_cached` into a
-    process-specific cache directory.
```
Comment (CharlelieLrt) [inline physicsnemo/utils/checkpoint.py:1372]: Similar comment here: the docstring is changed by the function did not. Is it because the docstring was wrong/outdated?

**Review by CharlelieLrt — APPROVED** [review]
I just had a few minor questions, so approving for checkpointing utilities (I did not review other files). One general comments is that these checkpointing utilities are getting more and more complicated so it would be good to simplify them at some point. Following previous discussions with @pzharrington my understanding was that this would be much simpler with FSDP2, but it seems it has instead gotten more complicated.

File: physicsnemo/utils/checkpoint.py (hunk 5 above)
Comment (negin513) [reply physicsnemo/utils/checkpoint.py:108]: Yes! That is the plan. FSDP1 is deprecated from PyTorch -- which is why I commented out FSDP1 related ones for when we want to drop this.

Comment (negin513) [reply physicsnemo/utils/checkpoint.py:108]: Correct! We should drop FSDP1 at some point. Not sure when to be implemented. I raised it up in the Group meeting. 
https://github.com/NVIDIA/physicsnemo/pull/1671#pullrequestreview-4439734875

I tried adding some helper docstrings so it helps when we want to remove FSDP1.

File: test/diffusion/test_domain_parallel_sampling.py (hunk 4 above)
Comment (negin513) [reply test/diffusion/test_domain_parallel_sampling.py:74]: shard_dim is a required argument with no default.  

https://github.com/NVIDIA/physicsnemo/blob/5693df2a8fdf2b8a98739b17f50f3d5cdf283fc1/physicsnemo/diffusion/noise_schedulers/domain_parallel.py#L40

File: physicsnemo/utils/checkpoint.py [hunk 8]
```diff
@@ -1364,12 +1520,9 @@ def get_checkpoint_dir(base_dir: Path | str, model_name: str) -> str:
 
 
 def _cache_if_needed(path: str) -> str:
-    r"""Return a local path for ``path``, downloading to cache if remote.
+    """
+    For remote fsspec URIs, downloads to a local cache and returns the local path.
```
Comment (negin513) [inline physicsnemo/utils/checkpoint.py]:
```suggestion

    Return a local path for ``path``, downloading to cache if remote.
    
    For the ``"file"`` protocol the path is returned unchanged.  For remote
    protocols the file is fetched via
    :func:`~physicsnemo.core.filesystem._download_cached` into a
    process-specific cache directory.
```

File: physicsnemo/utils/checkpoint.py (hunk 6 above)
Comment (negin513) [reply physicsnemo/utils/checkpoint.py:588]: The behvaior is still there, I just removed it as part of the docstring cleanup. I'll add it back to keep the docs accurate.

File: physicsnemo/utils/checkpoint.py (hunk 7 above)
Comment (negin513) [reply physicsnemo/utils/checkpoint.py:1372]: reverted back to the exact same wordings.

**Comment by negin513** [comment]
/ok to test [d877eb7](https://github.com/NVIDIA/physicsnemo/pull/1671/commits/d877eb794c55fe4948db3dffb4c081edfb57e6c2)
